### PR TITLE
Convert cudf::repeat to use device_uvector instead of device_vector

### PR DESCRIPTION
**Pull request by harrism**
Converts `cudf::repeat` to use device_uvector as much as possible, and converts to using indexalator for computing offsets from repeat counts. 

Also adds a benchmark for `cudf::repeat`.

Contributes to #7287 

Performance is improved:

```
Comparing /home/mharris/rapids/cudf/cpp/build/repeat_before.json to /home/mharris/rapids/cudf/cpp/build/repeat_after.json
Benchmark                                                       Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------
Repeat/double_nulls/1024/1/manual_time                       -0.1681         -0.1361             0             0             0             0
Repeat/double_nulls/4096/1/manual_time                       -0.1663         -0.1329             0             0             0             0
Repeat/double_nulls/32768/1/manual_time                      -0.1702         -0.1372             0             0             0             0
Repeat/double_nulls/262144/1/manual_time                     -0.1347         -0.1159             0             0             0             0
Repeat/double_nulls/2097152/1/manual_time                    -0.1053         -0.1010             0             0             0             0
Repeat/double_nulls/16777216/1/manual_time                   -0.0931         -0.0926             3             3             3             3
Repeat/double_nulls/67108864/1/manual_time                   -0.0869         -0.0868            12            11            12            11
Repeat/double_nulls/1024/8/manual_time                       -0.0539         -0.0462             0             0             0             0
Repeat/double_nulls/4096/8/manual_time                       -0.0951         -0.0839             0             0             0             0
Repeat/double_nulls/32768/8/manual_time                      -0.0905         -0.0821             0             0             0             0
Repeat/double_nulls/262144/8/manual_time                     -0.0499         -0.0453             0             0             0             0
Repeat/double_nulls/2097152/8/manual_time                    -0.0322         -0.0325             1             1             1             1
Repeat/double_nulls/16777216/8/manual_time                   -0.0267         -0.0272            10             9            10             9
Repeat/double_nulls/67108864/8/manual_time                   -0.0276         -0.0267            39            38            39            38
Repeat/double_no_nulls/1024/1/manual_time                    -0.2183         -0.1590             0             0             0             0
Repeat/double_no_nulls/4096/1/manual_time                    -0.2308         -0.1711             0             0             0             0
Repeat/double_no_nulls/32768/1/manual_time                   -0.2374         -0.1820             0             0             0             0
Repeat/double_no_nulls/262144/1/manual_time                  -0.1886         -0.1571             0             0             0             0
Repeat/double_no_nulls/2097152/1/manual_time                 -0.1266         -0.1200             0             0             0             0
Repeat/double_no_nulls/16777216/1/manual_time                -0.1045         -0.1039             3             2             3             2
Repeat/double_no_nulls/67108864/1/manual_time                -0.0945         -0.0944            11            10            11            10
Repeat/double_no_nulls/1024/8/manual_time                    -0.0595         -0.0490             0             0             0             0
Repeat/double_no_nulls/4096/8/manual_time                    -0.1201         -0.1028             0             0             0             0
Repeat/double_no_nulls/32768/8/manual_time                   -0.1085         -0.0968             0             0             0             0
Repeat/double_no_nulls/262144/8/manual_time                  -0.0670         -0.0635             0             0             0             0
Repeat/double_no_nulls/2097152/8/manual_time                 -0.0405         -0.0400             1             1             1             1
Repeat/double_no_nulls/16777216/8/manual_time                -0.0352         -0.0354             8             8             8             8
Repeat/double_no_nulls/67108864/8/manual_time                -0.0295         -0.0295            32            31            32            31
```

### CODE REVIEW

File: cpp/src/filling/repeat.cu [hunk 1]
```diff
@@ -72,36 +72,24 @@ struct count_accessor {
...
-  std::enable_if_t<std::is_integral<T>::value, rmm::device_vector<cudf::size_type>> operator()(
-    bool check_count, rmm::cuda_stream_view stream)
+  std::enable_if_t<std::is_integral<T>::value, void> operator()(rmm::cuda_stream_view stream)
   {
     // static_cast is necessary due to bool
-    if (check_count && static_cast<int64_t>(std::numeric_limits<T>::max()) >
-                         std::numeric_limits<cudf::size_type>::max()) {
-      auto max = thrust::reduce(p_column->begin<T>(), p_column->end<T>(), 0, thrust::maximum<T>());
+    if (static_cast<int64_t>(std::numeric_limits<cudf::size_type>::max()) >
+        std::numeric_limits<cudf::size_type>::max()) {
```
Comment (ttnghia) [inline cpp/src/filling/repeat.cu:83]: Sorry, I'm confused. Why are you comparing `static_cast<int64_t>(some_number) > some_number`? From my view point, `std::numeric_limits<cudf::size_type>::max()` always returns a `size_type` number (int32_t). You cast it to `int64_t` and do the comparison with it?

Comment (harrism) [reply cpp/src/filling/repeat.cu:83]: This was the existing check. I think it really should be a static_assert.

Comment (harrism) [reply cpp/src/filling/repeat.cu:83]: Can you approve now?

**Review by ttnghia — APPROVED** [review]

**Review by robertmaynard — APPROVED** [review]

**Review by cwharris — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**